File: server/src/main/scala/producer/message_generator/value_generator/ValueGenerator.scala
package producer.message_generator.value_generator

import com.tools.teal.pulsar.ui.producer.v1.producer as pb
import producer.message_generator.data_generators.bytes.BytesGenerator
import producer.message_generator.data_generators.json_generator.JsonGenerator
import org.graalvm.polyglot.Context
import org.apache.pulsar.client.api.TypedMessageBuilder
import org.apache.pulsar.common.schema.SchemaInfo
import _root_.producer.jsonToValue

case class ValueGenerator(
    generator: BytesGenerator | JsonGenerator | ValueFromTopicSchemaGenerator
):
    def generateMut(
        builder: TypedMessageBuilder[Array[Byte]],
        polyglotContext: Context,
        schemaInfo: Option[SchemaInfo]
    ): Unit =
        generator match
            case v: BytesGenerator =>
                val bytes = v.generate(polyglotContext)
                builder.value(bytes)

            case v: JsonGenerator =>
                val json = v.generate(polyglotContext)

                schemaInfo match
                    case None =>
                        builder.value(json.getBytes)
                    case Some(si) =>
                        jsonToValue(si, json.getBytes) match
                            case Left(e)  => throw new Exception(s"Failed to convert json to value: $e")
                            case Right(v) => builder.value(v)

            case v: ValueFromTopicSchemaGenerator => ???

object ValueGenerator:
    def fromPb(v: pb.ValueGenerator): ValueGenerator =
        val generator = v.generator match
            case pb.ValueGenerator.Generator.GeneratorFromBytes(v)            => BytesGenerator.fromPb(v)
            case pb.ValueGenerator.Generator.GeneratorFromJson(v)             => JsonGenerator.fromPb(v)
            case pb.ValueGenerator.Generator.GeneratorValueFromTopicSchema(v) => ValueFromTopicSchemaGenerator.fromPb(v)
            case _                                                            => throw new Exception("Unknown value generator type")

        ValueGenerator(
            generator = generator
        )

    def toPb(v: ValueGenerator): pb.ValueGenerator =
        val generator = v.generator match
            case v: BytesGenerator                => pb.ValueGenerator.Generator.GeneratorFromBytes(BytesGenerator.toPb(v))
            case v: JsonGenerator                 => pb.ValueGenerator.Generator.GeneratorFromJson(JsonGenerator.toPb(v))
            case v: ValueFromTopicSchemaGenerator => pb.ValueGenerator.Generator.GeneratorValueFromTopicSchema(ValueFromTopicSchemaGenerator.toPb(v))

        pb.ValueGenerator(
            generator = generator
        )

File: ui/components/ui/ProducerSession/producer-task/message-generator/value-generator/value-generator.ts
import * as pb from '../../../../../../grpc-web/tools/teal/pulsar/ui/producer/v1/producer_pb';
import { BytesGenerator, bytesGeneratorFromPb, bytesGeneratorToPb } from '../data-generators/bytes/bytes-generator';
import { JsonGenerator, jsonGeneratorFromPb, jsonGeneratorToPb } from '../data-generators/json-generator/json-generator';
import { ValueFromTopicSchemaGenerator, valueFromTopicSchemaGeneratorFromPb, valueFromTopicSchemaGeneratorToPb } from './value-from-topic-schema-generator';

export type ValueGenerator = {
  type: 'value-generator',
  generator: {
    type: "from-bytes",
    generator: BytesGenerator
  } | {
    type: "from-json",
    generator: JsonGenerator
  } | {
    type: "value-from-topic-schema",
    generator: ValueFromTopicSchemaGenerator
  }
};

export function valueGeneratorFromPb(v: pb.ValueGenerator): ValueGenerator {
  let generator: ValueGenerator['generator'];
  switch (v.getGeneratorCase()) {
    case pb.ValueGenerator.GeneratorCase.GENERATOR_FROM_BYTES:
      generator = {
        type: "from-bytes",
        generator: bytesGeneratorFromPb(v.getGeneratorFromBytes()!)
      };
      break;
    case pb.ValueGenerator.GeneratorCase.GENERATOR_FROM_JSON:
      generator = {
        type: "from-json",
        generator: jsonGeneratorFromPb(v.getGeneratorFromJson()!)
      };
      break;
    case pb.ValueGenerator.GeneratorCase.GENERATOR_VALUE_FROM_TOPIC_SCHEMA:
      generator = {
        type: "value-from-topic-schema",
        generator: valueFromTopicSchemaGeneratorFromPb(v.getGeneratorValueFromTopicSchema()!)
      };
      break;
    default:
      throw new Error(`Unknown value generator: ${v.getGeneratorCase()}`);
  }

  return { type: 'value-generator', generator };
}

export function valueGeneratorToPb(v: ValueGenerator): pb.ValueGenerator {
  const message = new pb.ValueGenerator();

  switch (v.generator.type) {
    case 'from-bytes': {
      message.setGeneratorFromBytes(bytesGeneratorToPb(v.generator.generator));
      break;
    }
    case 'from-json': {
      message.setGeneratorFromJson(jsonGeneratorToPb(v.generator.generator));
      break;
    }
    case 'value-from-topic-schema': {
      message.setGeneratorValueFromTopicSchema(valueFromTopicSchemaGeneratorToPb(v.generator.generator));
      break;
    }
  }

  return message;
}
